### PR TITLE
encryption: support multiple cloud providers

### PR DESCRIPTION
**Pull request by gregwebs**
Signed-off-by: Greg Weber <greg@pingcap.com>

### What problem does this PR solve?

Problem Summary:

TiKV needs to support different KMS encryption providers. For example, issue #8906 

### What is changed and how it works?

What's Changed:

The KMS backend now has a KmsProvider which defaults to AwsKms.
It is now straightforward to write a different backend.

Some associated refactorings were made:
* Simplify KmsBackend internal state management
* Creating a backend does not happen inside Arc::new()
* Reduce code pollution of testing where `#[cfg(test)]` was used
* Add a Debug trait to backends
* Add some newtypes for safety
* Make some error messages more detailed
* Fix KMS was marked as not send but can change how it uses the APIs instead

### Related changes

None, cherry-pick to a release branch may be needed in the future.

### Check List 

Tests 

- Unit test
- Integration test (tested manually with file backend and KMS backend to see that encryption is enabled and that keys are rotated)

### Release note 

Support alternative KMS encryption backends

### CODE REVIEW

File: components/external_storage/src/s3.rs
Comment (yiwu-arbug) [inline components/external_storage/src/s3.rs:37]: return error instead of panic.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:72]: nit: just have `region`, `endpoint` and `current_key_id` as fields, no need to define a `KmsClientDebug` type and not printing runtime?

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:104]: return error instead of panic as much as possible.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:226]:
```suggestion
pub struct KmsDataKey {
```
Want to distinguish it with `kvproto::encryptionpb::DataKey`.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:116]: rename the type `EncryptedRef` to `CiphertextKey`, if you want to use the newtype ideom. Definition of the type should be moved to this file, as it is KMS specific.

Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:116]: It would be nice to also define a new type for the plaintext key.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:240]: Since `KmsProvider` is `Sync`, we don't need a mutex to guard it.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:228]: should avoid the plaintext key being printed.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:226]:
```suggestion
pub struct DataKeyPair {
```

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:245]: Shall we keep the existing behavior to lazily initialize the kms data key? In case preexisting encryption metadata presents, we don't need this extra `generate_data_key` call.

File: components/encryption/src/master_key/kms.rs
Comment (gregwebs) [reply components/encryption/src/master_key/kms.rs:240]: The Mutex is to avoid requiring an `&mut` for `KmsBackend` in the `decrypt` function.

File: components/encryption/src/config.rs
Comment (yiwu-arbug) [reply components/encryption/src/config.rs:54]: yes, do you mind help me update the macro?

Comment (gregwebs) [reply components/encryption/src/config.rs:54]: I think it would be better to make this change separately. I have experience doing the same thing in our go code and always found that the solution to these issues is just to delete any custom aws credentials code and use what the SDK provides by default.

File: components/encryption/src/manager/mod.rs
Comment (gregwebs) [reply components/encryption/src/manager/mod.rs:245]: In that case can we check to see that the pre-existing plaintext file exists on the file system?

File: components/encryption/src/config.rs
Comment (yiwu-arbug) [reply components/encryption/src/config.rs:54]: "use what the SDK provides by default` is implemented in the `new_client!` macro. We just need a variant of the macro that don't look at the custom credential but just use the default credentials.

File: components/encryption/src/crypter.rs
Comment (yiwu-arbug) [inline components/encryption/src/crypter.rs:243]:
```suggestion
// PlainKey is a newtype used to mark a vector a plaintext key.
```

File: components/encryption/src/manager/mod.rs
Comment (yiwu-arbug) [reply components/encryption/src/manager/mod.rs:245]: we always assume the source file exists on the file system when `link_file` is called. If the file entry exists in the file dictionary, we assume the file is encrypted, otherwise unencrypted.

File: components/encryption/src/config.rs
Comment (yiwu-arbug) [reply components/encryption/src/config.rs:54]: the explicit credentials is not being used by anyone for the KMS flow. It is only used for S3 for BR.

File: components/encryption/src/crypter.rs
Comment (yiwu-arbug) [inline components/encryption/src/crypter.rs:204]: how about `EncryptedKey`, to be consistent with `PlainKey`?

File: components/encryption/src/manager/mod.rs
Comment (yiwu-arbug) [reply components/encryption/src/manager/mod.rs:632]: then why we need to refactor `get_file`? `get_file_exists` looks redundant. Again if the file entry doesn't exist in file dictionary, it could mean the file does not exist, or the file exists but not encrypted. In the key manager we don't care about if the file exists on the file system. We return file info with method = plaintext and let the caller decide what to do.

Comment (yiwu-arbug) [reply components/encryption/src/manager/mod.rs:632]: Or in another word, we always assume the caller of `get_file` knows the file exists.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:72]: do you have a sample output for the `runtime`?

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:122]: well, in this context it is the encrypted data key. it is more clear if the type name is specific.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:251]: why you want to avoid `&mut`?

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:323]: Add comment to say the KMS data key is rotated every time on restart? This is different from previous behavior but I think it is a good change.

File: components/encryption/src/crypter.rs
Comment (gregwebs) [reply components/encryption/src/crypter.rs:204]: PlainKey differentiates a key from other unencrypted data. `Encrypted` can be used to mark different types of data (a key or a payload) as encrypted. I could see this developing into `Key<T>` where `T` could be `Encrypted` or not.

File: components/encryption/src/master_key/kms.rs
Comment (gregwebs) [reply components/encryption/src/master_key/kms.rs:251]: The other backends and the interface don't need `&mut`, and `encrypt` does not need it, so I was hesitant to change the interface to `&mut` just for `decrypt`: to me that interface definition would look odd. I suppose I can add some comments saying that it is mutable for the purposes of caching internal state.

File: components/encryption/src/crypter.rs
Comment (yiwu-arbug) [reply components/encryption/src/crypter.rs:204]: I don't see the code being more clear after introducing this type. It doesn't serve any functional purpose, and don't make the logic more clear.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:251]: I think you can remove `&mut` from `KmsProvider::generate_data_key`, then you don't need a mutex for the `kms_provider`.

Comment (gregwebs) [reply components/encryption/src/master_key/kms.rs:251]: In `generate_data_key` `retry()` requires `runtime` to be mutable.

File: components/encryption/src/crypter.rs
Comment (gregwebs) [reply components/encryption/src/crypter.rs:204]: I think I agree with you because we already have an EncryptedContent type in use.

File: components/encryption/src/manager/mod.rs
Comment (gregwebs) [reply components/encryption/src/manager/mod.rs:245]: It is strange to have to assume so much. I am wondering if this can be improved. However, it seems that some of this behavior is forced by the RocksDB encryption implementation?

Comment (yiwu-arbug) [reply components/encryption/src/manager/mod.rs:245]: not really. we make use of some of original rocksdb encryption implementation, but the key manager logic are new. The design of the key manager and the encryption module is that it has no knowledge of the actual file system and only care about encryption metadata. Interaction with file system should be handled in the higher level, i.e. the work-in-progress `file_system` crate.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:272]: Not familiar with tokio, but maybe `runtime` should not be member of `AwsKms`, instead being created for each invocation of `decrypt`?

Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:272]: cc @overvenus

**Comment by yiwu-arbug** [comment]
cc @overvenus for logic related to #7236 fix.

File: components/encryption/src/crypter.rs
Comment (overvenus) [inline components/encryption/src/crypter.rs:225]: Could you also sanitize `std::fmt::Display`?

File: components/encryption/src/encrypted_file/mod.rs
Comment (overvenus) [reply components/encryption/src/encrypted_file/mod.rs:73]: Panic here might be simpler than propagating errors, this error is fatal, callers may panic anyway.

File: components/external_storage/src/s3.rs
Comment (overvenus) [inline components/external_storage/src/s3.rs:33]: Without this phantom, `S3Storage` is error-prone, it may cause panic if it is used in other threads.

Comment (gregwebs) [reply components/external_storage/src/s3.rs:33]: From my reading the underlying github issues, the problem is from using the global dispatcher. The code below here is changed to instead create our own dispatcher. Let me know if you have any specific tests for this that I could run.

File: components/encryption/src/manager/mod.rs
Comment (gregwebs) [reply components/encryption/src/manager/mod.rs:245]: Can the caller check for this case before calling the function? Then we don't need any assumptions.

File: components/encryption/src/master_key/kms.rs
Comment (gregwebs) [reply components/encryption/src/master_key/kms.rs:271]: Runtimes are generally designed to be persistent. I suspect we should not be creating a runtime but instead passing one in from a top level. That will still require a mutable runtime. However, in either case if we assume all KmsProvider needs the tokio runtime we can pass it in to the decrypt_data_key function.

Comment (gregwebs) [reply components/encryption/src/master_key/kms.rs:271]: Okay, runtime is now in State and passed into the backend, so the backend is now immutable and does not need a Mutex.

File: components/encryption/src/config.rs
Comment (gregwebs) [reply components/encryption/src/config.rs:54]: I think we are in agreement. I will make a separate PR for this though.

File: components/encryption/src/encrypted_file/mod.rs
Comment (gregwebs) [reply components/encryption/src/encrypted_file/mod.rs:73]: It was already performing a panic. I need to limit the scope of my improvements in this large PR. There are unwraps everywhere in our code base.

File: components/encryption/src/config.rs
Comment (yiwu-arbug) [reply components/encryption/src/config.rs:54]: okay

File: components/encryption/src/manager/mod.rs
Comment (yiwu-arbug) [reply components/encryption/src/manager/mod.rs:245]: That's not an assumption. The point is the encryption crate should not aware of file system stuff, and just focus on encryption metadata.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:226]: maybe we don't even want the encrypted key to be printed?

File: components/encryption/src/crypter.rs
Comment (yiwu-arbug) [inline components/encryption/src/crypter.rs:201]: typo: valid

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [inline components/encryption/src/master_key/kms.rs:337]: scoping the mutex lock for readablity:
```
{
    let mut state = self.state.lock().unwrap();
    // the rest of code that uses the state
}
```

File: components/encryption/src/master_key/kms.rs
Comment (gregwebs) [reply components/encryption/src/master_key/kms.rs:226]: In theory it is safe to publish encrypted data publicly, so I am not worried about the security. I think we need some way to identify the data key when debugging.

File: components/encryption/src/encrypted_file/mod.rs
Comment (yiwu-arbug) [reply components/encryption/src/encrypted_file/mod.rs:73]: well, i actually prefer propagate the error to upper layer, which prints the error message of encryption failure, before panicking. I hope we can reduce the unnecessary panics one by one, but since the panic exists previously, it is okay not to update it.

File: components/encryption/src/master_key/kms.rs
Comment (gregwebs) [reply components/encryption/src/master_key/kms.rs:337]: it's the entire rest of the function so the behavior and the compiler analysis is exactly the same. Is this better for developers to read the code? It is also possible it will be confusing to someone reading the code since they don't have any effect.

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs]: cool

File: components/external_storage/src/s3.rs
Comment (yiwu-arbug) [reply components/external_storage/src/s3.rs:33]: @gregwebs I mainly have two questions for the issue, do you mind explain a bit?
1. how does `tokio::runtime::Builder::basic_scheduler()` work? From the document it says the runtime will execute tasks on "current thread". What if I create a basic runtime on thread T1, send it over to thread T2, and execute some task? Does T1 or T2 count for "current thread"?
2. how does creating a new dispatcher for each `AmsKms` and `S3Storage` instances (which are `Sync + Send`, so they allow people to send them to different threads)  fixes the issue?

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:226]: okay

File: components/encryption/src/master_key/kms.rs
Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:337]: It is a common practices to scope the lock guard for readability. It make it crystal clear how far is the critical session.

Comment (yiwu-arbug) [reply components/encryption/src/master_key/kms.rs:337]: for example in an later iteration people add more unrelated logic at the end of this function, and may forget to unlock the mutex. explicit scoping avoids that.

File: components/external_storage/src/s3.rs
Comment (gregwebs) [reply components/external_storage/src/s3.rs:33]: 1) I don't know. The behavior after this PR is merged should be the same as previous.

2) My reading of the issue is that there is a global dispatcher that is initialized with lazy_static and it turns out it is not actually safe to use it. But instead you can just create a dispatcher at runtime yourself.

Comment (gregwebs) [reply components/external_storage/src/s3.rs:33]: This is the default shared client: https://github.com/rusoto/rusoto/blob/d59d716f097f64a713b3bbcd5fc7b84e210647e4/rusoto/core/src/client.rs#L16
This is how it is used: https://github.com/rusoto/rusoto/search?q=Client%3A%3Ashared

I don't know specifically what the problem is, but I can easily believe that this would cause problems.

File: components/encryption/src/manager/mod.rs
Comment (gregwebs) [reply components/encryption/src/manager/mod.rs:245]: We talked about this and agreed that in a separate PR we can change this from `Ok(()` to `None` and return an `Option` type.

File: components/external_storage/src/s3.rs
Comment (gregwebs) [reply components/external_storage/src/s3.rs:33]: Since S3 does not need this change I will keep S3 as is with a note about how this is unnecessary now. I will remove the phantom from KMS because it shouldn't be necessary and even if there were a race condition in hyper it would be difficult for KMS to hit it now that it uses its own client since it very rarely makes http calls.

File: components/encryption/src/master_key/kms.rs
Comment (gregwebs) [reply components/encryption/src/master_key/kms.rs]: okay, the prior behavior is restored now

**Comment by yiwu-arbug** [comment]
Waiting for testing and @overvenus 's review for the tokio runtime related logic, before approving.

**Comment by gregwebs** [comment]
@yiwu-arbug I went ahead and removed the static credentials as per your suggestion. They were interfering with my testing.

File: components/external_storage/src/s3.rs
Comment (gregwebs) [reply components/external_storage/src/s3.rs:33]: [This comment](https://github.com/hyperium/hyper/issues/2112#issuecomment-630857751) indicates that avoiding the lazy_static client (connected to a global dispatcher) as the code does now fixes the issue.

File: components/encryption/src/manager/mod.rs
Comment (overvenus) [inline components/encryption/src/manager/mod.rs:237]: Why returning `Result<Option<()>>`? Is it possible to replace it with `Result<bool>`?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: components/encryption/src/manager/mod.rs
Comment (gregwebs) [reply components/encryption/src/manager/mod.rs:237]: Both represent two states. The problem with using `bool` is knowing what it means. For `is_empty` it is obvious, but in cases like these it is not. The file apis here are being changed to return an `Option` with `None` meaning the operation did not occur because the file did not exist. So `Option<T>` is the consistent choice for these APIs and in this case `Option<()>` will be understandable to someone familiar with the APIs.

File: components/sst_importer/src/util.rs
Comment (gregwebs) [reply components/sst_importer/src/util.rs:72]: I changed the behavior of link_file back to how it was previously, so I should remove this change now

Comment (gregwebs) [reply components/sst_importer/src/util.rs]: This is changed back now

**Comment by overvenus** [comment]
LGTM (including S3 client change)

**Comment by overvenus** [comment]
LGTM

**Comment by gregwebs** [comment]
I rebased the PR to 2 commits. I did a diff of the rebase and what was previously pushed for the branch to ensure they were the same.

**Comment by hunterlxt** [comment]
Sorry late for response, the PR is large so I read several times.

**Review by hunterlxt — APPROVED** [review]
/lgtm

**Review by yiwu-arbug — APPROVED** [review]
LGTM

**Pull request merged**
